File: Algorithms/Search/cpp/binary_search.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int binarySearch(int arr[], int n, int x) {
	int found = -1;
	int low = 0, high = n-1;

    while(low <= high) {
    	int mid = (low + high)/2;

    	if(arr[mid] < x) {
    		// search right tree
    		low = mid + 1;
    	}
    	else if(x > arr[mid]) {
    		// search right tree
    		high = mid - 1;
    	}
        else {
            found = mid;
            break;
        }
    }

    return found;
}

int main() {
	int n;
    // input size of array
    cin >> n;

    int arr[n];    
    // input array elements
    for(int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    int x;
    // element to be found
    cin >> x;

    // sort array
    sort(arr, arr+n);

    int found = binarySearch(arr, n, x);

	if(found >= 0) {
        cout << "Element found!";
    }
    else {
        cout << "Element not found!";
    }

    return 0;
}

File: Algorithms/Search/python/binary_search.py
def Binary_Search(Arr, find):
    """
    type Arr: list[int]
    type find: int
    """
    
    left = 0
    right = len(Arr)-1
    
    while(left <= right):
        mid = (left + right)//2
        
        if( Arr[mid] == find):
            return mid
        elif(Arr[mid] < find):
            left = mid + 1
        else:
            right = mid - 1
    
    return -1
    
if __name__ == "__main__":
    Arr = [0,8,31,46,79,93,950]
    find = 950
    loc = Binary_Search(Arr, find)
    if(loc == -1):
        print("Value not found in Array")
    else:
        print("Value found at index", loc)
